File: C++/problem2-2-1.cpp
/* @file                problem2-1-1.cpp
 * @author          zhaoqicheng(zhaoqicheng2022@163.com)
 * @brief              BGR2GRAY and BGR2HSV
 * @details
 * @version         0.1.0
 * @date              2022-4-12
 * @copyright    2022-4-12 by zhaoqicheng.All rights reserved.
 */
//RGB转HSV

#include <iostream>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

void RgbToHsv(Mat image_rgb, Mat &image_hsv)
{
  //创建一张三通道的HSV图像
  image_hsv = Mat::zeros(image_rgb.size(), CV_8UC3);
  //遍历实现操作
  for (int i = 0; i < image_rgb.rows; i++)
  {
    Vec3b *p0 = image_rgb.ptr<Vec3b>(i);   //B=p[0]  G=p[1]  R=p[2]
    Vec3b *p1 = image_hsv.ptr<Vec3b>(i);   //B=p[0]  G=p[1]  R=p[2]

    for (int j = 0; j < image_rgb.cols; j++)
    {
      //将图像任意坐标点的BGR分别转换到0-1之间
      float B = p0[j][0] / 255.0;
      float G = p0[j][1] / 255.0;
      float R = p0[j][2] / 255.0;

      float V = (float)std::max({ B, G, R });     //B/G/R
      float vmin = (float)std::min({ B, G, R });
      float diff = V - vmin;

      float S, H;
      S = diff / (float)(fabs(V) + FLT_EPSILON);
      diff = (float)(60.0 / (diff + FLT_EPSILON));

      if (V == B) { //V=B
        H = 240.0 + (R - G) * diff;
      }
      else if (V == G) {  //V=G
              H = 120.0 + (B - R) * diff;
      }
      else if (V == R) {  //V=R
        H = (G - B) * diff;
      }
     //如果如果计算得到的H值小于0，将该值再加上360，得到最终的H值
      H = (H < 0.0) ? (H + 360.0) : H;

      p1[j][0] = (uchar)(H / 2);
      p1[j][1] = (uchar)(S * 255);
      p1[j][2] = (uchar)(V * 255);
    }
  }
}

//RGB转GRAY
void RgbToGray(const Mat &image_rgb, Mat &image_gray) {
    if (!image_rgb.data || image_rgb.channels() != 3) {
        return;
    }
    //创建一张单通道的灰度图像
    image_gray = Mat::zeros(image_rgb.size(), CV_8UC1);
    //取出存储图像像素的数组的指针
    uchar *pointImage = image_rgb.data;
    uchar *pointImageGray = image_gray.data;
    //取出图像每行所占的字节数
    size_t stepImage = image_rgb.step;
    size_t stepImageGray = image_gray.step;
    for (int i = 0; i < image_gray.rows; i++) {
        for (int j = 0; j < image_gray.cols; j++) {
            pointImageGray[i*stepImageGray + j] =
                           (uchar)(0.114*pointImage[i*stepImage + 3 * j] + 0.587*pointImage[i*stepImage + 3 * j + 1]+0.299*pointImage[i*stepImage + 3 * j + 2]);
        }
    }
}


int main() {
    Mat image = imread("/home/jichuan-ftp/image/unchanged/lena.jpg");
    Mat gray,hsv;
    RgbToHsv(image, hsv);
    RgbToGray(image, gray) ;
    imshow("RGB",image);
    imshow("GRAY",gray);
    imshow("HSV",hsv);

    waitKey(0);
    destroyAllWindows();

    return 0;
}

File: 蟒/problem2-2-1.py
# @file		    problem2-2-1.py
# @author      	ZhaoQicheng(zhaoqicheng2022@163.com/love-Tom)
# @brief        RGB2GRAY/RGB2HSV
# @detail       自行设计算法实现RGB转GRAY功能与RGB转HSV功能，分别输出原图，灰度图和HSV图，不得调用现成函数
# @version 	    0.1.0
# @date 		2022-4-18
# @copyright	2022-4-18 by zhaoqicheng.All rights reserved. 
  
import cv2 as cv
import numpy as np

#函数定义
#RGB转GRAY函数
def RgbToGray(image):
    row, col, channel = image.shape  #读取图片的属性(行，列，通道数)
    gray = np.zeros((row, col))  #创建灰度图
    #遍历灰度图进行转换操作
    for r in range(row):
        for l in range(col):
            gray[r, l] = 0.299 * image[r, l, 0] + 0.587 * image[r, l, 1] + 0.114 * image[r, l, 2]  #转换公式
    
    cv.imshow("GRAY",gray.astype("uint8"))  #强制转换为整型
#RGB转HSV函数
def RgbToHsv(image):
    #读取RGB图像的属性
    row = image.shape[0]
    col = image.shape[1]
    #分别创建H,S,V图像
    H = np.zeros((row,col),np.float32)
    S = np.zeros((row,col),np.float32)
    V = np.zeros((row,col),np.float32)
    #分离RGB图像的三个通道
    r,g,b = cv.split(image)
    #将r,g,b转换到0-1之间
    r, g, b = r/255.0, g/255.0, b/255.0
    #遍历进行转换操作
    for i in range(0, row):
        for j in range(0, col):
            big = max((b[i, j], g[i, j], r[i, j]))
            small = min((b[i, j], g[i, j], r[i, j]))
            V[i, j] = big  #将最大值赋值给V
            if V[i, j] == 0:
                S[i, j] = 0  #最大值为0，S赋值为0
            else:
                S[i, j] = (V[i, j] - small) / V[i, j]
            if big == small:  #最大值与最小值相等，H赋0
                H[i, j] = 0  
            elif V[i, j] == r[i, j]:
                if g[i, j] >= b[i, j]:
                    H[i, j] = (60 * ((g[i, j]) - b[i, j]) / (V[i, j] - small))     #最大值等于r且g>=b
                else:
                    H[i, j] = (60 * ((g[i, j]) - b[i, j]) / (V[i, j] - small))+360 #最大值等于r且g<b
            elif V[i, j] == g[i, j]:
                H[i, j] = 60 * ((b[i, j]) - r[i, j]) / (V[i, j] - small) + 120     #最大值等于g
            elif V[i, j] == b[i, j]:
                H[i, j] = 60 * ((r[i, j]) - g[i, j]) / (V[i, j] - small) + 240     #最大值等于b
            H[i,j] = H[i,j] / 2
    #合并之前分离的H,S,V通道
    hsv = cv.merge([H,S,V])
    cv.imshow("HSV",hsv)


#函数调用
image = cv.imread("D:\photo\lena.jpg")  #读取图像
gray = RgbToGray(image)                 #生成灰度图
RgbToHsv(image)                         #生成HSV图
cv.imshow("RGB",image)                  #展示RGB图

cv.waitKey(0)
cv.destroyAllWindows()